### DISCUSSION TITLE
GTX 10 support?

### DISCUSSION THREAD

**Discussion by cracksoff** [terminatorul/NvStrapsReBar#39] [General]
Will there be support for GTX10?

**Comment by UnidentifiedTag** [comment]
Unfortunately not for now, this may change

**Comment by terminatorul** [comment]
Technically the tool will create a large bar for Pascal cards, but then the NVIDIA driver for Windows will crash or reset the system (creating a boot loop).

@UnidentifiedTag  I do not see how this may change, do you have more information ?

**Reply by UnidentifiedTag** [reply]
While Nvidia won’t open up the drivers there have been leaks of source code for these drivers, someone could use that to painfully peace together a working driver with this option turned on or added for 10 series, as of now I guess there hasn’t been a reason for anyone to look in to this or attempt it.

**Reply by mirh** [reply]
I'm reading that on linux BAR works but "driver does not pick up the new BAR size". 
Did you try with [the patch](https://github.com/JuliaComputing/nvidia-driver-pcie-rebar) that was being used before they [added](https://github.com/NVIDIA/open-gpu-kernel-modules/pull/3) official support (only in the Turing+ module though AFAIU)?

**Reply by terminatorul** [reply]
No, I didn't try the patch and I didn't try the driver on Linux. I never wanted to the install the proprietary driver. 

If someone like you @mirh can try it and the patch still works, it will be great news for sure !

**Reply by slapjd** [reply]
As far as I could tell, that patch does the same or a very similar process to what the official support does for kernel-open.

You can copy and paste `nv_resize_pcie_bars` from kernel-open to the proprietary module (kernel) before installing the driver (assuming you're using dkms) and it'll run, but it still doesn't work. It fails to find any possible ReBAR sizes (`pci_rebar_get_possible_sizes` kernel function) and then fails.

I *think* that kernel function is looking for the actual ReBAR PCIe capability which doesn't seem to exist on Pascal cards even with the UEFI patch, unless I'm doing something wrong.

**Reply by mirh** [reply]
https://patchwork.kernel.org/project/linux-pci/cover/20201218174011.340514-1-s.miroshnichenko@yadro.com/
There was also this patch floating around (which has a really nice [presentation](https://lpc.events/event/7/contributions/847/) too), which is said to make it more likely for the nvidia rebar patch to succeed finding free address space.

**Reply by ericnaustin97-cell** [reply]
With support ending for the pascal cards does that change anything?

**Comment by terminatorul** [comment]
The people at [nvidia-patch](https://github.com/keylase/nvidia-patch/) project apparently have a way to patch NVIDIA drivers to enable more NVENC encoders.

Not related to ReBAR, but patching the NVIDIA driver is also the first step needed to fix ReBAR for the Pascal driver. I don't have reverse engineering experience except to scratch the surface, but if someone (like you ??) can raise a ticket on that project, maybe the authors can help here as well, with finding the faulty instruction in the Pascal driver that triggeres the blue screen BSOD ?

**Comment by tai** [comment]
While I'm also stuck with nvidia driver recognizing only 256MB on Quadro P400 (Pascal with 2GB VRAM), I have some input which might make reverse engineering effort somewhat easier: PCIe BAR1 resize can be done without UEFI hack if you are on Linux.

Here's how you can do it:
```
# lspci -Dnnn | grep P400
0000:10:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP107GL [Quadro P400] [10de:1cb3] (rev a1)
# echo 0000:10:00.0 > /sys/bus/pci/drivers/nvidia/unbind
# pcimem /sys/bus/pci/devices/0000\:10\:00.0/resource0 0x10100C w 0x82413000
# cat /proc/cmdline
...omitted usual options for clarity... intel_iommu=off amd_iommu=off pci=nocrs,realloc
# kexec -l /vmlinuz --initrd=/initrd.img --reuse-cmdline
# kexec -e
```

This updates PSTRAPS1 register on GPU and boot into another Linux, causing another PCI bus initialization and resource allocation with updated BAR1 size. All other resources , like memory regions allocated to other devices/bridges, are also adjusted accordingly.

Please note that the value you write to resource0 (BAR0 region) depends on initial PSTRAPS1 register value.

```
# pcimem /sys/bus/pci/devices/0000\:10\:00.0/resource0 0x101000 w*8
/sys/bus/pci/devices/0000:10:00.0/resource0 opened.
Target offset is 0x101000, page size is 4096
mmap(0, 4096, 0x3, 0x1, 3, 0x101000)
PCI Memory mapped to address 0x7f862baea000.
0x101000: 0x00400080  <- PSTRAPS0 register value
0x101004: 0xBADF5040
...
0x10100C: 0x02013000   <- PSTRAPS1 register value
0x101010: 0xBADF5040
```

For my case (Quadro P400), it's 0x02013000. As described by [envytools](https://envytools.readthedocs.io/en/latest/hw/io/pstraps.html) for G80+ families, sum of PSTRAPS0[15:14] and PSTRAPS1[22:20] determines BAR1 size. Also, highest bit (1<<31) needs to be set to enable writeback. Hence the following command:

```
(gdb) p/x (0x02013000|1<<31|1<<22)
$1 = 0x82413000

# pcimem /sys/bus/pci/devices/0000\:10\:00.0/resource0 0x10100C w 0x82413000
/sys/bus/pci/devices/0000:10:00.0/resource0 opened.
Target offset is 0x10100c, page size is 4096
mmap(0, 4096, 0x3, 0x1, 3, 0x10100c)
PCI Memory mapped to address 0x7f5aeb806000.
0x10100C: 0x02013000
Written 0x82413000; readback 0x82413000
```

In above, I have set PSTRAPS1[22:22] bit and this gave me 4GB BAR1 size after kexec-ed reboot:

```
# lspci -vvvs 10:00.0 | grep Region
        Region 0: Memory at f6000000 (32-bit, non-prefetchable) [size=16M]
        Region 1: Memory at 200000000 (64-bit, prefetchable) [size=4G]
        Region 3: Memory at 180000000 (64-bit, prefetchable) [size=32M]
        Region 5: I/O ports at f000 [size=128]
```

This still doesn't solve the issue of nvidia driver (and nvidia-smi -q) reporting only 256MB BAR1 size even after Linux's PCI subsystem has allocated larger BAR1 size. But I hope this (hopefully) easier method let other people hack into the issue!

**Reply by tai** [reply]
After I recompiled nvidia driver with a debug version, I noticed that the driver actually reports enlarged 4GB BAR1 size in its debug output:

```
# nvidia-smi -q
...
    BAR1 Memory Usage
        Total                             : 256 MiB
        Used                              : 2 MiB
        Free                              : 254 MiB
...
# dmesg | grep BAR1
[   16.192372] NVRM: PCI:0000:10:00.0 (10de:1cb3): BAR1 @ 0x200000000 (4096MB)
[   16.486769] NVRM: PCI:0000:10:00.0 (10de:1cb3): BAR1 @ 0x200000000 (4096MB)
```

I'm now wondering what if we just override nvmlDeviceGetBAR1MemoryInfo() API in libnvidia-ml.so and just force application to work with enlarged BAR1 size...

**Reply by mirh** [reply]
That was noted [here](https://github.com/xCuri0/ReBarUEFI/discussions/89#discussioncomment-7596828) yes (and FWIW [this](https://github.com/CFSworks/nvml_fix) seems to go along your idea).
I thought applications didn't have to do anything to support rebar though, and it was just a driver backend thing to handle better?
